tools/Makefile.shared

```diff
@@ -12,10 +12,62 @@
 #*   special exception on linking described in the file LICENSE.          *
 #*                                                                        *
 #**************************************************************************
-
+MAKEFLAGS := -r -R
 include ../config/Makefile
-CAMLRUN ?= ../boot/ocamlrun
-CAMLYACC ?= ../boot/ocamlyacc
+ifndef CAMLRUN
+CAMLRUN := ../boot/ocamlrun
+endif
+ifndef CAMLYACC
+CAMLYACC := ../boot/ocamlyacc
+endif
+DESTDIR ?=
+
+# Setup GNU make variables storing per-target source and target,
+# a list of installed tools, and a function to quote a filename for
+# the shell.
+override target=''""'$(subst ','\'',$@)'#'#')'#
+override source=''""'$(subst ','\'',$<)'#'#')'#
+override shellquote=''""'$(subst ','\'',$1)'#'#')'#
+override installed_tools := ocamldep ocamlprof ocamlcp ocamloptp \
+                   ocamlmktop ocamlmklib ocamlobjinfo
+
+install_files :=
+define byte2native
+$(patsubst %.cmo,%.cmx,$(patsubst %.cma,%.cmxa,$1))
+endef
+
+# $1 = target, $2 = OCaml object dependencies, $3 = other dependencies
+#
+# There is a lot of subtle code here.  The multiple layers of expansion
+# are due to `make`'s eval() function, which evaluates the string
+# passed to it as a makefile fragment.  So it is crucial that variables
+# not get expanded too many times.
+define byte_and_opt_
+# This check is defensive programming
+$(and $(filter-out 1,$(words $1)),$(error \
+   cannot build file with whitespace in name))
+$1: $3 $2
+	$$(CAMLC) -compat-32 $$(LINKFLAGS) -o $$(target) $2
```

Comment:
`-compat-32` is only needed for `ocamldep`. You shouldn't add it to every target without at least discussing the change.